File: ext/microsandbox/src/agent.rs
//! Raw agent client: `Microsandbox::Native::AgentClient`.
//!
//! Mirrors `sdk/python/src/agent.rs`. Wraps the core `AgentBridge` — the
//! FFI-shaped, bytes-in/bytes-out façade over a sandbox's agentd relay socket.
//! Frames are moved as raw CBOR bodies; (de)serialization stays in Ruby. Streams
//! are referenced by opaque `u64` handles so the Ruby layer never owns a tokio
//! receiver. Every call runs on the shared tokio runtime with the GVL released.

use std::sync::Arc;
use std::time::Duration;

use magnus::{function, method, prelude::*, Error, RHash, RModule, RString, Ruby};
use microsandbox::agent::AgentClient as CoreAgentClient;
use microsandbox::{AgentBridge, BridgeFrame, MicrosandboxError};

use crate::error;
use crate::runtime::{block_on, ruby};

/// Map an agent-client error onto the Ruby exception hierarchy (via the core
/// `MicrosandboxError::AgentClient` wrapper, exactly like the Python binding).
fn to_ruby_agent(err: microsandbox::AgentClientError) -> Error {
    error::to_ruby(MicrosandboxError::AgentClient(err))
}

#[magnus::wrap(class = "Microsandbox::Native::AgentClient", free_immediately, size)]
pub struct AgentClient {
    inner: Arc<AgentBridge>,
}

impl AgentClient {
    fn from_bridge(bridge: AgentBridge) -> Self {
        Self {
            inner: Arc::new(bridge),
        }
    }

    //----------------------------------------------------------------------
    // Connection (singleton methods)
    //----------------------------------------------------------------------

    /// Connect to a running sandbox by name. `timeout` is optional seconds.
    fn connect_sandbox(name: String, timeout: Option<f64>) -> Result<AgentClient, Error> {
        let bridge = match dur(timeout)? {
            Some(t) => block_on(AgentBridge::connect_sandbox_with_timeout(&name, t)),
            None => block_on(AgentBridge::connect_sandbox(&name)),
        }
        .map_err(to_ruby_agent)?;
        Ok(AgentClient::from_bridge(bridge))
    }

    /// Connect to an agentd relay socket by path. `timeout` is optional seconds.
    fn connect_path(path: String, timeout: Option<f64>) -> Result<AgentClient, Error> {
        let bridge = match dur(timeout)? {
            Some(t) => block_on(AgentBridge::connect_path_with_timeout(&path, t)),
            None => block_on(AgentBridge::connect_path(&path)),
        }
        .map_err(to_ruby_agent)?;
        Ok(AgentClient::from_bridge(bridge))
    }

    /// Resolve a sandbox's agent relay socket path without connecting.
    fn socket_path(name: String) -> Result<String, Error> {
        let path = CoreAgentClient::socket_path(&name).map_err(error::to_ruby)?;
        Ok(path.to_string_lossy().into_owned())
    }

    //----------------------------------------------------------------------
    // Instance methods
    //----------------------------------------------------------------------

    /// Send one frame and await a single response frame ({id, flags, body}).
    fn request(&self, flags: u8, body: RString) -> Result<RHash, Error> {
        let body = unsafe { body.as_slice() }.to_vec();
        let inner = Arc::clone(&self.inner);
        let frame =
            block_on(async move { inner.request(flags, body).await }).map_err(to_ruby_agent)?;
        Ok(frame_to_hash(frame))
    }

    /// Open a streaming session; returns {id, handle}.
    fn stream_open(&self, flags: u8, body: RString) -> Result<RHash, Error> {
        let body = unsafe { body.as_slice() }.to_vec();
        let inner = Arc::clone(&self.inner);
        let (id, handle) =
            block_on(async move { inner.stream_open(flags, body).await }).map_err(to_ruby_agent)?;
        let hash = ruby().hash_new();
        hash.aset("id", id)?;
        hash.aset("handle", handle)?;
        Ok(hash)
    }

    /// Pull the next frame from a stream; nil at end-of-stream.
    fn stream_next(&self, handle: u64) -> Result<Option<RHash>, Error> {
        let inner = Arc::clone(&self.inner);
        let frame =
            block_on(async move { inner.stream_next(handle).await }).map_err(to_ruby_agent)?;
        Ok(frame.map(frame_to_hash))
    }

    /// Close a stream handle. Idempotent.
    fn stream_close(&self, handle: u64) -> Result<(), Error> {
        let inner = Arc::clone(&self.inner);
        block_on(async move { inner.stream_close(handle).await });
        Ok(())
    }

    /// Send a follow-up frame on an existing correlation id.
    fn send(&self, id: u32, flags: u8, body: RString) -> Result<(), Error> {
        let body = unsafe { body.as_slice() }.to_vec();
        let inner = Arc::clone(&self.inner);
        block_on(async move { inner.send(id, flags, body).await }).map_err(to_ruby_agent)
    }

    /// Cached handshake `core.ready` frame body bytes (CBOR).
    fn ready_bytes(&self) -> Result<RString, Error> {
        let bytes = self.inner.ready_bytes().map_err(to_ruby_agent)?;
        Ok(ruby().str_from_slice(&bytes))
    }

    /// Close the connection. Idempotent.
    fn close(&self) -> Result<(), Error> {
        let inner = Arc::clone(&self.inner);
        block_on(async move { inner.close().await });
        Ok(())
    }
}

/// Convert an optional `timeout` (seconds) into an optional `Duration`,
/// mirroring the Python SDK's `timeout_duration`:
///   - absent (`nil`) → `None` → the core's default handshake timeout
///   - `0` → an explicit zero deadline (fail fast), *not* "use the default"
///   - negative or non-finite (NaN/Inf) → a caller error (rather than being
///     silently swallowed into the default)
fn dur(timeout: Option<f64>) -> Result<Option<Duration>, Error> {
    match timeout {
        None => Ok(None),
        Some(t) if t.is_finite() && t >= 0.0 => Ok(Some(Duration::from_secs_f64(t))),
        Some(t) => Err(error::base_error(format!(
            "timeout must be a non-negative, finite number of seconds (got {t})"
        ))),
    }
}

/// Shape a `BridgeFrame` into a Ruby Hash. `body` is binary (ASCII-8BIT).
fn frame_to_hash(frame: BridgeFrame) -> RHash {
    let r = ruby();
    let hash = r.hash_new();
    let _ = hash.aset("id", frame.id);
    let _ = hash.aset("flags", frame.flags);
    let _ = hash.aset("body", r.str_from_slice(&frame.body));
    hash
}

pub fn define(ruby: &Ruby, native: &RModule) -> Result<(), Error> {
    let class = native.define_class("AgentClient", ruby.class_object())?;

    class.define_singleton_method(
        "connect_sandbox",
        function!(AgentClient::connect_sandbox, 2),
    )?;
    class.define_singleton_method("connect_path", function!(AgentClient::connect_path, 2))?;
    class.define_singleton_method("socket_path", function!(AgentClient::socket_path, 1))?;

    class.define_method("request", method!(AgentClient::request, 2))?;
    class.define_method("stream_open", method!(AgentClient::stream_open, 2))?;
    class.define_method("stream_next", method!(AgentClient::stream_next, 1))?;
    class.define_method("stream_close", method!(AgentClient::stream_close, 1))?;
    class.define_method("send", method!(AgentClient::send, 3))?;
    class.define_method("ready_bytes", method!(AgentClient::ready_bytes, 0))?;
    class.define_method("close", method!(AgentClient::close, 0))?;

    Ok(())
}

File: lib/microsandbox/agent.rb
# frozen_string_literal: true

module Microsandbox
  # A single raw agent-protocol frame: a correlation id, flag bits, and a
  # CBOR-encoded body. Decode {#body} with a CBOR library of your choice.
  class AgentFrame
    # @return [Integer] correlation id from the frame header
    attr_reader :id
    # @return [Integer] flag bits (see {AgentClient::FLAG_TERMINAL} etc.)
    attr_reader :flags
    # @return [String] raw CBOR body bytes (ASCII-8BIT)
    attr_reader :body

    def initialize(data)
      @id = data["id"]
      @flags = data["flags"]
      @body = data["body"]
    end

    # @return [Boolean] last frame of a stream
    def terminal? = (@flags & AgentClient::FLAG_TERMINAL) != 0

    # @return [Boolean] opens a new session
    def session_start? = (@flags & AgentClient::FLAG_SESSION_START) != 0

    # @return [Boolean] connection-shutdown signal
    def shutdown? = (@flags & AgentClient::FLAG_SHUTDOWN) != 0

    def inspect
      "#<Microsandbox::AgentFrame id=#{@id} flags=0x#{@flags.to_s(16)} body=#{@body&.bytesize}B>"
    end
  end

  # An open raw agent stream, from {AgentClient#stream}. {Enumerable}: iterate to
  # consume {AgentFrame}s until the stream ends (the terminal frame is delivered,
  # then iteration stops). Mirrors the official SDKs' `AgentStream`.
  #
  # @note **Single-pass, forward-only, single-consumer.** `each` drains a
  #   one-shot native channel — not rewindable, iterate once from a single
  #   thread; a second pass or a post-drain combinator yields nothing.
  #
  # @example
  #   stream = client.stream(0, request_body)
  #   stream.each { |frame| handle(frame) }
  class AgentStream
    include Enumerable

    # @return [Integer] the protocol correlation id (pass to {AgentClient#send_frame})
    attr_reader :id

    def initialize(native, id, handle)
      @native = native
      @id = id
      @handle = handle
    end

    # Pull the next frame, or nil at end-of-stream.
    # @return [AgentFrame, nil]
    def recv
      frame = @native.stream_next(@handle)
      frame && AgentFrame.new(frame)
    end

    # @yieldparam frame [AgentFrame]
    # @return [self, Enumerator]
    def each
      return enum_for(:each) unless block_given?

      while (frame = recv)
        yield frame
      end
      self
    end

    # Close the stream and release its handle. Idempotent.
    # @return [nil]
    def close
      @native.stream_close(@handle)
      nil
    end
  end

  # A low-level **raw agent client** — the byte-level transport to a sandbox's
  # `agentd` over its relay socket. This is the rawest tier of the SDK: it moves
  # CBOR-encoded protocol frames in and out; encoding/decoding the bodies is up
  # to you. Most users want {Sandbox#exec}/{Sandbox#fs} instead; reach for this
  # to drive `agentd` protocol features the high-level API does not expose, or to
  # bridge the relay socket to another transport (see {socket_path}).
  #
  # Mirrors the `AgentClient` in the official Python/Node/Go SDKs.
  #
  # @example one-shot request/response
  #   Microsandbox::AgentClient.connect_sandbox("my-box") do |client|
  #     frame = client.request(0, cbor_encoded_request)
  #     handle(frame.body)
  #   end
  class AgentClient
    # Frame flag bits (mirror the protocol constants in the other SDKs).
    FLAG_TERMINAL = 0b0000_0001
    FLAG_SESSION_START = 0b0000_0010
    FLAG_SHUTDOWN = 0b0000_0100

    class << self
      # Connect to a running sandbox by name (max 128 UTF-8 bytes). With a block,
      # the client is yielded and closed when the block returns.
      # @param name [String]
      # @param timeout [Numeric, nil] handshake timeout in seconds. nil (the
      #   default) uses the core default (~10s); 0 fails fast (an immediate
      #   deadline); a negative or non-finite value raises {Error}.
      # @yieldparam client [AgentClient]
      # @return [AgentClient, Object]
      def connect_sandbox(name, timeout: nil, &block)
        wrap(Native::AgentClient.connect_sandbox(name.to_s, timeout && Float(timeout)), &block)
      end

      # Connect to an agentd relay socket by path. See {connect_sandbox}.
      # @return [AgentClient, Object]
      def connect_path(path, timeout: nil, &block)
        wrap(Native::AgentClient.connect_path(path.to_s, timeout && Float(timeout)), &block)
      end

      # Resolve a sandbox's agent relay socket path **without** connecting — the
      # same path {connect_sandbox} would dial. Useful for bridging the socket to
      # another byte transport. The sandbox need not be running.
      # @return [String]
      def socket_path(name)
        Native::AgentClient.socket_path(name.to_s)
      end

      private

      def wrap(native, &block)
        client = new(native)
        return client unless block

        begin
          block.call(client)
        ensure
          client.close
        end
      end
    end

    def initialize(native)
      @native = native
    end

    # Send one frame and await a single response frame.
    # @param flags [Integer] frame flag bits
    # @param body [String] CBOR-encoded body bytes
    # @return [AgentFrame]
    def request(flags, body)
      AgentFrame.new(@native.request(Integer(flags), body.to_s))
    end

    # Open a streaming session.
    # @param flags [Integer]
    # @param body [String]
    # @return [AgentStream]
    def stream(flags, body)
      opened = @native.stream_open(Integer(flags), body.to_s)
      AgentStream.new(@native, opened["id"], opened["handle"])
    end

    # Send a follow-up frame on an existing correlation id (e.g. a stream's
    # {AgentStream#id}). Named +send_frame+ rather than +send+ so it does not
    # shadow Ruby's `Object#send`. Maps to the protocol "send" in the other SDKs.
    # @return [nil]
    def send_frame(id, flags, body)
      @native.send(Integer(id), Integer(flags), body.to_s)
      nil
    end

    # The cached handshake `core.ready` frame body (CBOR bytes).
    # @return [String]
    def ready_bytes
      @native.ready_bytes
    end

    # Close the connection. Idempotent.
    # @return [nil]
    def close
      @native.close
      nil
    end
  end
end
